### DISCUSSION TITLE
PCM usage in multi-threaded application

### DISCUSSION THREAD

**Discussion by VenturaDelMonte** [opcm/pcm#212] [Q&A] [answered]
Hi all, I hope this is the right place to ask my question. I would like to profile some sections of a multithreaded application.  Each code section is executed within one thread on a partition of the input. What i want to do is to profile each code section that runs on a specific core. I am a bit confused regarding the multi-threaded model of PCM. Can one PCM instance be shared across threads? Would that be expensive, e.g., is there a lock?Is this an intended way of using PCM or should one use it in a dedicated thread that samples stats every N milliseconds (e.g., like in the provided examples).
In short, my code would look like:
```
for (int i = 0; i < numThreads; ++i) {
     // set thread affinity to core i and numa
     auto* input = data[i];
     // start sampling events using PCM for core i
     // do some critical, coarse-grained work
     // stop sampling and report stats
     // do some other work
     for (int j = 0; j  < elementsInInput; ++j) {
          // start profiling again using PCM for core i
          // do some fine-grained work
          // stop profiling and report stats 
     }
}
```
I am aware that this could also be done via Vtune but I would be interested in using pcm only.

EDIT: is `allowMultipleInstance()` meant to solve the above problem or is it to allow multiple PCM processes at a time?

**Comment by opcm** [comment]
what you should do is the following:
```
program pcm once

for (int i = 0; i < numThreads; ++i) {
     // set thread affinity to core i and numa
     auto* input = data[i];
     auto before = getCoreCounterState(i);
     // do some critical, coarse-grained work
     auto after = getCoreCounterState(i);
     // print metrics of interest, for example getL3CacheMisses(before, after);
     // do some other work
     for (int j = 0; j  < elementsInInput; ++j) {
         auto before = getCoreCounterState(i);
         // do some critical, coarse-grained work
         auto after = getCoreCounterState(i);
         // print metrics of interest, for example getL3CacheMisses(before, after);
     }
}
```
getCoreCounterState has only per-core/thread locks inside, so it should not matter for you

**Comment by VenturaDelMonte** [comment]
hi @opcm thank you very much for your help. Much appreciated. It works!

Please, feel free to close the issue or let me know if i can help making this snippet available for future use.

**Comment by VenturaDelMonte** [comment]
@opcm a followup question, if i may. Would it be possible to configure a different set of custom counters on each core? If that is not possible, would you recommend reprogramming the PCM instance with different counters every N units of time or should the PCM instance be untouched after `program(..)` is called? Thanks in adv. for your attention.

**Comment by opcm** [comment]
you can change the set of monitored events for all cores altogether: https://github.com/opcm/pcm/blob/4f326d4b45b29fe996e9a6b60034eeee28a6e8f6/pcm-core.cpp#L89

I expect reprogramming should work.

**Comment by VenturaDelMonte** [comment, marked as answer]
@opcm thank you! i confirm that reprogramming on-the-fly the PCM instance works.
